src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -142,7 +143,25 @@ public void WriteResponseHeaders(int statusCode, string ReasonPhrase, HttpRespon
                     return;
                 }
 
-                _frameWriter.WriteResponseHeaders(_streamId, statusCode, responseHeaders);
+                // If the responseHeaders will be written to the final frame then set
+                // END_STREAM on the HEADERS frame. This avoids the need to write an
+                // empty DATA frame with END_STREAM
+                //
+                // The headers will be the final frame if there is no content and there
+                // is no trailing HEADERS frame.
+                Http2HeadersFrameFlags http2HeadersFrame;
+
+                if (appCompleted && responseHeaders.ContentLength == 0 && (_stream.Trailers == null || _stream.Trailers.Count == 0))
```

Comment:
Why rely on ContentLength 0? Wouldn't it be better to check if the app completed without ever starting to write to the body?

Reply:
This is needed for the case where someone has written to a span on the underlying PipeReader, then set the content length.

> Wouldn't it be better to check if the app completed without ever starting to write to the body?

How do I do that?

Reply:
I guess you mean _startedWritingDataFrames. I've changed to use that.